### ISSUE TITLE
Homebrew support - roll back or let me disable?

### ISSUE THREAD

**Issue by fuzzy76** [mangerlahn/Latest#288]
Regarding the homebrew support: thanks, I hate it.

It shows updates for apps that doesn't update via homebrew, but via their own update mechanisms. Which means that I have to open all of the actual apps to update, which is different from how all updates in Latest.app has worked up until now.. This feels terrible, tbh.

To add to the problem, settings seems disabled, so I can't find any way to turn it off.

**Comment by mangerlahn** [comment]
All of these apps would update via Homebrew, but in-app updating for homebrew-based updates is not yet implemented due to time constraints on my side. I figured it was time to release what I had now. In-app updating for these apps is still planned, I want to do that next.
For the time being, you can always ignore updates for apps you don't care about at the moment.

**Comment by streamtree** [comment]
I'll add a comment to #285, but I don't think there's any need to check all entries.

**Comment by fuzzy76** [comment]
The problem is that I have most apps on my computer installed using homebrew. So the update list in latest is completely drowned in apps I don't want to / can't update using latest. This makes ignoring pretty hard, since I have to search the list for actual updateable apps every time...

**Comment by fuzzy76** [comment]
Just to clarify: my issue with this is not false positives.

**Comment by ycardon** [comment]
I agree with @fuzzy76, an option to disable the Homebrew upgrade feature until it's working fine would be welcome. I've got dozens of false positive : review each of them every time is painful and prevents me from using your wonderful app. 

Regarding the false positives, there seems to be an interference with the Homebrew feature to not upgrade Casks that upgrade themselves (`auto_updates true` in the formula, forceable by `brew upgrade --cask --greedy`)

**Comment by fuzzy76** [comment]
I want Latest.app to only show me apps that I can update from within Latest.app using a single click. Instead I get to launch every single app to update manually.

**Comment by Stanzilla** [comment]
Yeah should be an option imo, or maybe a separate category in the list. Also I don't think the way you are currently detecting these new updates is going to scale well, its going to need constant maintenance of false positives etc

**Comment by fuzzy76** [comment]
> The idea was to be able to check for updates outside the scope of Mac App Store and Sparkle

That sounds fine. But do you really think it's a good idea to check for and show updates that Latest can't update?

**Comment by mangerlahn** [comment]
That's how Latest functioned for years with App Store and Sparkle apps. No-one ever cared so I figured it would be fine as a first step. I see how this confusing now that some apps offer in-app updating. Ill see what I can do to add updating for Brew as well.

**Comment by Stanzilla** [comment]
> That's how Latest functioned for years with App Store and Sparkle apps. No-one ever cared so I figured it would be fine as a first step. I see how this confusing now that some apps offer in-app updating. Ill see what I can do to add updating for Brew as well.

well that is fine if the app actually has a way to manually update but some don't, see Google Drive for example 
![Google Chrome Beta-2023-12-13-17 21 15](https://github.com/mangerlahn/Latest/assets/75278/c2b58f46-2c1d-47f4-8afe-7902bfad69db)

**Comment by fuzzy76** [comment]
> Ill see what I can do to add updating for Brew as well.

If you intend to update those that actually updateable by homebrew by doing a `brew upgrade` on them, that's fine.

The part to notice (and it seems like you haven't) is that casks with `auto_updates true` explicitly tells homebrew that they would like to update themselves. In those cases you should 1) not show them at all in Latest and 2) *not* attempt to update them.

If you attempt to "update anyway" by doing a reinstall of the cask you risk some casks overwriting their settings (because they might do a uninstall in the process), or break stuff in other ways. They might even have other reasons for asking brew to leave them alone, which you don't know about. Please do not do this.

**Comment by mangerlahn** [comment]
In case the user has installed homebrew, using the commands provided by brew is desired, as this ensures Latest does not mess with brew.

`auto_updates true` is a good point, regardless of whether brew is installed or not. That value is also set on apps that update via Sparkle, which Latest has been updating for quite some time now. In other cases it is probably always a good idea to open the app, so that it can update itself through whatever mechanism it uses. For apps with `auto_updates false`, `brew upgrade` would be ideal. If brew is not installed, Latest could download the linked binary, however there is still some complexity with the different types of binaries that get downloaded this way. (dmg vs. zip. vs pkg and so on..)

However, I don't get why Latest should not show these updates. Latest is an app to list which apps have updates available. That is reason number 1 this app exists. All the other stuff is just bonus.

**Comment by fuzzy76** [comment]
For me (and probably a lot of other users), Latest is an app to list **and update** apps with updates available. That's probably why some people have different expectations of what that list should contain.

My use flow includes launching Latest.app every morning and updating all apps. If that require launching large monstrosities like LibreOffice, Jetbrains Datagrip, Firefox, Pycharm and such even when I don't intend to use them that day, I'd rather remove Latest.app from my routine all together.

Ofcourse, all of this could be fixed with a setting. "Don't show apps that Latest.app can't update".

**Comment by fuzzy76** [comment]
> If brew is not installed, Latest could download the linked binary, however there is still some complexity with the different types of binaries that get downloaded this way. (dmg vs. zip. vs pkg and so on..)

I would advise against that. Some formulas have custom install scripts for a reason. You could end up breaking the install. If an app isn't installed using homebrew, I see no reason it should be updated using homebrew formulas.

**Comment by mangerlahn** [comment]
> Ofcourse, all of this could be fixed with a setting. "Don't show apps that Latest.app can't update".

Thats a good differentiation and I am happy to add that as a setting. Until now, it was not clear to me what your issue was here.

> Some formulas have custom install scripts for a reason. You could end up breaking the install. If an app isn't installed using homebrew, I see no reason it should be updated using homebrew formulas.
There are many different cases here for every app that is checked: 
- Homebrew not installed → Open app to update / (try to download something?, optional, might be too hard with many edge cases)
- Homebrew installed:
    - App installed via brew → brew upgrade
    - App not installed via brew → Handle as if homebrew was not installed

Did I miss anything?

That said, don't expect any of this soon. It clearly needs some thought and time to implement. If anyone wants to contribute, I'd highly appreciate.

**Comment by fuzzy76** [comment]
No that sounds right :)

**Comment by tzimpel** [comment]
> * Homebrew not installed → Open app to update / (try to download something?, optional, might be too hard with many edge cases)
> * Homebrew installed:
>   
>   * App installed via brew → brew upgrade
>   * App not installed via brew → Handle as if homebrew was not installed
> 
> Did I miss anything?

Yes, you did, actually.

* Apps installed via homebrew (as cask) are only updatable via `brew upgrade` when they don't provide their own in-app update mechanism. All apps with their own update mechanism have `auto_updates true` set in the cask. So you also need to handle these apps as if home-brew was not installed.
* The same is true for apps that use `version :latest`, even if they don't have an update mechanism of their own, as homebrew would not be able to know which version is installed or if an update is available.

The only way to update them via homebrew would be to use `brew reinstall`, which isn't a good way to update these apps.

See https://docs.brew.sh/FAQ#why-arent-some-apps-included-during-brew-upgrade for a more detailed description

**Comment by fuzzy76** [comment]
> > Ofcourse, all of this could be fixed with a setting. "Don't show apps that Latest.app can't update".

> Thats a good differentiation and I am happy to add that as a setting. Until now, it was not clear to me what your issue was here.

For now I've downgraded to 0.9 on all my Macs , but would really like to be able to use the latest releases. Do you want a separate issue for this feature request, or should this issue be regarded as it?

**Comment by du-song** [comment]
Thank you for making this nice piece of software. to me the brew based update detection works quite well, all I hope here is to update them.

to avoid some edge cases for some users, maybe you can offer it as an opt-in to begin with. 

Is this the issue about a feature request of update via home-brew for those can? A bit confused after reading the conversation above. Happy to create a new issue if that make things clear, and beta testing.
